src/parser/warlock/demonology/modules/pets/Pets.js

```diff
@@ -0,0 +1,425 @@
+    pet.currentEnergy = newEnergy;
+    if (pet.currentEnergy === 0) {
+      pet.realDespawn = event.timestamp;
+      test && this.log('Despawning Wild Imp', pet);
+    }
+  }
+
+  // to update player position more accurately (based on DistanceMoved)
+  // important, since Wild Imp summons uses player position as default (not entirely accurate, as they're spawned around player, not exactly on top of it, but that's as close as I'm gonna get)
+  // needed for Implosion - there's a possibility that Wild Imps don't cast anything between their 'summon' and Implosion, therefore I wouldn't get their position
+
+  on_toPlayer_damage(event) {
+    this._updatePlayerPosition(event);
+  }
+
+  on_toPlayer_energize(event) {
+    this._updatePlayerPosition(event);
+  }
+
+  on_toPlayer_heal(event) {
+    this._updatePlayerPosition(event);
+  }
+
+  on_toPlayer_absorbed(event) {
+    this._updatePlayerPosition(event);
+  }
+
+  // API
+
+  get currentPets() {
+    return this._getPets();
+  }
+
+  getPetDamage(id, isGuid = true) {
+    // isGuid = true, because it's more convenient to call this with getPetDamage(PETS.SOME_PET.guid)
+    // because you know what you're looking for (pet IDs change, GUIDs don't)
+    const guid = isGuid ? id : this._toGuid(id);
+    if (!this.petDamage[guid]) {
+      debug && this.error(`this.getPetDamage() called with nonexistant ${isGuid ? 'gu' : ''}id ${id}`);
+      return 0;
+    }
+    return Object.values(this.petDamage[guid].instances).reduce((total, current) => total + current, 0);
+  }
+
+  get permanentPetDamage() {
+    // haven't observed any real rule for permanent pet guids except the fact, that they're the longest (other pet guids are either 5 or 6 digits)
+    let total = 0;
+    Object.entries(this.petDamage).filter(([guid]) => guid.length > 6).forEach(([guid, pet]) => {
+      total += Object.values(pet.instances).reduce((total, current) => total + current, 0);
+    });
+    return total;
+  }
+
+  getPetCount(timestamp = this.owner.currentTimestamp, petId = null) {
+    return this._getPets(timestamp).filter(pet => petId ? pet.id === petId : true).length;
+  }
+
+  get petsBySummonAbility() {
+    return this.petTimeline.reduce((obj, pet) => {
+      const key = pet.summonedBy || 'unknown';
+      const spellName = pet.summonedBy ? SPELLS[pet.summonedBy].name : 'unknown';
+      obj[key] = obj[key] || { spellName, pets: [] };
+      obj[key].pets.push(pet);
+      return obj;
+    }, {});
+  }
+
+  // HELPER METHODS
+
+  _getPets(timestamp = this.owner.currentTimestamp) {
+    return this.petTimeline.filter(pet => pet.spawn <= timestamp && timestamp <= (pet.realDespawn || pet.expectedDespawn));
+  }
+
+  _getPetFromTimeline(id, instance) {
+    return this.petTimeline.find(pet => pet.id === id && pet.instance === instance);
+  }
+
+  _isPermanentPet(guid) {
```

Comment:
Isn't it possible to have an array of permanent pet guids?

Reply:
From what I found across several logs, the guids seem random even for the same pet type (primary pet for Demonology is Felguard), whereas for the summoned pets, they're the same.

Reply:
Hmm maybe it's related to your permanent pets name? Maybe you could use a blacklist instead then, assume it's a permanent pet if it doesn't match one of the summoned pets' guid.